src/python/Utils/wmcoreDTools.py

```diff
@@ -0,0 +1,706 @@
+            continue
+        elif entry["type"] == "thread":
+            threadPids.append(int(entry["pid"]))
+
+    # Check if initial threads are running.
+    # NOTE: The list threadPids is fetched from the threads.json file
+    #       as it has been constructed at Daemon startup time. Any threads spawn later during
+    #       the component's lifetime are not tracked by threads.json file, which means
+    #       they will always fall either into orphanThreads or lostThreads. That's why the
+    #       results from this function cannot be used for components like AgentWatchdog,
+    #       where every timer has its own thread and those are regularly restarted.
+    #       Such mechanism always results in non-constant and nonzero values in the orphanThreads and
+    #       lostThreads fields of the pidTree, which are changing following the child threads life cycle.
+    process = psutil.Process(int(pid))
+    currThreads = set([thread.id for thread in process.threads()])
+    startupThreads = set(threadPids)
+    # Remove the parent pid from the list of currently running threads
+    try:
+        currThreads.remove(int(pid))
+    except ValueError or KeyError:
+        pass
+
+    runningThreads = currThreads & startupThreads
+    orphanThreads = currThreads - startupThreads
+    lostThreads = startupThreads - currThreads
+
+    # Adding an exception for AgentWatchdog when it comes to building the sets of threads
+    if component == 'AgentWatchdog':
+        runningThreads = currThreads
+        orphanThreads = set()
+        LostThreads = set()
+
+    # Output result
+    msg=""
+    runningMsg=""
+    orphanMsg=""
+    lostMsg=""
+    status = "running" if daemon.isAlive() else "not-running"
+    if status == "running":
+        runningMsg = f"with {len(runningThreads)} running threads: {runningThreads}"
+        if len(lostThreads) > 0:
+            status = f"{status}-partially"
+            lostMsg = f", {len(lostThreads)} lost threads: {lostThreads}"
+        if len(orphanThreads) > 0:
+            status = f"{status}-untracked"
+            orphanMsg = f", {len(orphanThreads)} untracked/zombie threads: {orphanThreads}"
+    msg = f"Component:{component} {pid} {status} {runningMsg} {lostMsg} {orphanMsg}"
+    if not quiet:
+        logging.info(msg)
+
+    pidTree['Parent'] = int(pid)
+    pidTree['RunningThreads'] = list(runningThreads)
+    pidTree['OrphanThreads'] = list(orphanThreads)
+    pidTree['LostThreads'] = list(lostThreads)
+    return pidTree
+
+def restart(config, componentsList=None, doLogCleanup=False, doDirCleanup=False):
+    """
+    _restart_
+
+    do a shutdown and startup again
+    :param configFile: Either path to the WMAgent configuration file or a WMCore.Configuration instance
+    :return:           int ExitCode - 0 in case of success, nonzero value otherwise
+    """
+    exitCode = 0
+    exitCode += shutdown(config, componentsList, doDirCleanup, doLogCleanup)
```

Comment:
I don't like sum of error codes since it will mislead of what actually happen. Instead I suggest to return a dict of error codes to clearly define their nature, e.g. `{"shutdown": <shutdown error code>, "startup": "<startup error code>", ...}`. Having such dictionary can be helpful to define proper set of action upon recovery or can give full information to viewer to know at which step system hickup.